sap-service-operator/src/test/java/com/sap/cloud/environment/servicebinding/FileSystemWatcherCacheTest.java

```diff
@@ -239,6 +243,69 @@ void loadWillIgnoreBindingWhenLoaderReturnsNull( @Nonnull @TempDir final Path ro
         verify(loader, times(1)).apply(eq(dir));
     }
 
+    @Test
+    @SuppressWarnings( "unchecked" )
+    void watchServiceIsCached()
+        throws Exception
+    {
+        final Function<Path, ServiceBinding> loader = (Function<Path, ServiceBinding>) mock(Function.class);
+        final FileSystem fileSystemA = mock(FileSystem.class);
+        final FileSystem fileSystemB = mock(FileSystem.class);
+        final WatchService watchService = mock(WatchService.class);
+
+        when(fileSystemA.newWatchService()).thenReturn(watchService);
+        when(fileSystemB.newWatchService()).thenReturn(watchService);
+
+        new FileSystemWatcherCache(loader, fileSystemA);
+        verify(fileSystemA, times(1)).newWatchService();
+
+        new FileSystemWatcherCache(loader, fileSystemA);
+        verify(fileSystemA, times(1)).newWatchService();
+
+        // using a different FileSystem will create a new WatchService
+        new FileSystemWatcherCache(loader, fileSystemB);
+        verify(fileSystemB, times(1)).newWatchService();
+    }
+
+    @Test
+    @SuppressWarnings( "unchecked" )
+    void finalizeCancelsWatchKeys()
+    {
+        final Function<Path, ServiceBinding> loader = (Function<Path, ServiceBinding>) mock(Function.class);
+        when(loader.apply(any())).thenReturn(null);
+
+        final WatchKey watchKey = mock(WatchKey.class);
+
+        final FileSystemWatcherCache sut = new FileSystemWatcherCache(loader);
+
+        sut.directoryWatchKeys.put(Paths.get("some", "imaginary", "path"), watchKey);
+
+        sut.finalize();
+
+        verify(watchKey, times(1)).cancel();
+        assertThat(sut.directoryWatchKeys).isEmpty();
+    }
+
+    @Disabled( "This test doesn't work reliably in the CI/CD pipeline. It can still be used for manual testing." )
+    @Test
+    @SuppressWarnings( "unchecked" )
+    void watchKeysAreCancelledOnGcRun()
+    {
+        final Function<Path, ServiceBinding> loader = (Function<Path, ServiceBinding>) mock(Function.class);
+        when(loader.apply(any())).thenReturn(null);
+
+        final WatchKey watchKey = mock(WatchKey.class);
+
+        FileSystemWatcherCache sut = new FileSystemWatcherCache(loader);
+
+        sut.directoryWatchKeys.put(Paths.get("some", "imaginary", "path"), watchKey);
+
+        sut = null;
+
+        System.gc();
+        verify(watchKey, times(1)).cancel();
```

Comment:
I think here it could make sense to directly call the `finalize` method. There is no guarantee that it will be called on the `System.gc()` call.
In the end you want to test that the implementation of `finalize` is correct, and not that/when the JVM is calling it.

Reply:
We already have a [test that checks whether the implementation of `finalize` is actually correct](https://github.com/SAP/btp-environment-variable-access/blob/fix/limit-watch-service-instantiations/sap-service-operator/src/test/java/com/sap/cloud/environment/servicebinding/FileSystemWatcherCacheTest.java#L270-L287).

This test here has the purpose of showing that `finalize` will also be called by the GC, but as timing might be hard to reproduce consistently, it is disabled (so that it won't jeopardize our CI/CD).